Various optimizations to generator, inference, and training

mikejhuang opened the pull request:
# Summary of changes
Implemented the following optimizations as described and benchmarked below
* OphysGenerator
  * batch 2D slicing
  * GPU caching of movie
    * full-movie
    * minimum frames required to generate batch
* Global optimizations
  * intel-tensorflow-avx512 for CPU inference optimization
  * mixed float16 mode for training and inference for devices with hardware support
* core_inference optimizations
  * [multiprocessing pool for inference implemented in separate PR](https://github.com/AllenInstitute/deepinterpolation/pull/93 )
# Optimizations
## Generator
Identified generator inefficiencies from the master branch include:  
* Reloading of the movie on each slice operation of the generator.
  * **Optimization:** Cache movie. This has been implemented by both Scott and Jerome in their own respective inference optimizations. 
* Slicing is performed for each frame to generate the model inputs and outputs for that frame. These are stored in larger batch objects for the input and output. Each slice operation on a large movie is expensive.
  * **Optimization:** Consolidate all slices across a batch into a single 2D slice. e.g. instead of batch[0]=arr[1,2,3], batch[1]=arr[4,5,6]. batch[2]=arr[7,8,9], do batch=arr[[1,2,3],[4,5,6],[7,8,9]]
* GPU inference is an I/O intensive operation as the large batch objects are transferred to GPU. However, given the sliding window nature of the batches, most data of a batch is redundant. 
  * **Optimization:** Copy the movie (or chunks of it that fit in memory) onto the GPU and generate batches by slicing from the movie on the GPU. [HPC nodes from 2019 and onwards](http://confluence.corp.alleninstitute.org/display/ITO/High+Performance+Computing) utilize NVIDIA A100 TPUs containing 40GB or 32GB of memory. This is large enough to fit entire movies on there.
* Normalization is performed with numpy on a single core and can take a non-trivial amount of time for large movies
  * **Optimization:** Use numba to parallelize the operation

## Inference
### CPU inference
* **Optimization:** [intel-tensorflow-avx512](https://pypi.org/project/intel-tensorflow-avx512/) Use a tensorflow build that is configured to utilize CPU instruction sets for optimized vector processing. Intel's AVX512, or also known as DL-Boost is present on Synapse and all [HPC nodes from 2019 and onwards](http://confluence.corp.alleninstitute.org/display/ITO/High+Performance+Computing). HPC Nodes from 2016 and 2017 support AVX and can benefit from a build configured to support those instructions.
* **Optimization:** Python multiprocessing instead of Tensorflow utilization of Intel Math Kernel Library for multithreading. Scott implemented the Python multiprocessing for tensorflow inference. I was initially dubious this would be faster since Python is infamous for its limitations with multiprocessing due to the [Global Interpreter Lock](https://realpython.com/python-gil/). Tensorflow, on the other hand, is implemented with an efficient C++ backend that utilizes [Intel Math Kernel Library](https://www.intel.com/content/dam/develop/external/us/en/documents/mkl-affinity.pdf) that is optimized at the level of machine code to perform multithreading for matrix operations.  However, with empirical tests, Python multiprocessing outperformed Tensorflow's native multithreading. 
  * **Caveat** The current implementation of multiprocessing does not allow the memory to be shared with all processes and hence the data is duplicated across the processes. For example, with a batch size of 16, the tensorflow multithreading uses 29G of memory, while the multiprocessing version uses 74G (includes cached movie with duplicated batches). However, this should not pose an issue on the HPC or Synapse with RAM sizes at 256. 

### GPU Inference
* **Optimization:** [float16](https://www.tensorflow.org/guide/mixed_precision) offers large speedups with minimal losses in model performance. NVIDIA GPUs or TPUs such as the A100 used in the HPC has native FP16 hardware support. Older GPUs such as the 1080Ti used in the nodes 2017 and earlier have limited FP16 support and will see less improvement in performance. 
* **Optimization:** GPU based generator as described in the generator section. This affects inference since the initial step of model.predict_on_batch is to convert the numpy array to a tensor on the GPU. This transfer of data takes up the bulk of the processing time while the GPU is sitting idle during this time. By removing this transfer step, the GPU will be spending most of its compute time performing inference.

# Methods
## Testbed
### CPU Workstation (Synapse)
* 16C/32T Intel Gold 6242 with AVX2, AVX512
* 256GB RAM

### GPU Workstation
* 16/32T AMD Ryzen 5950X with AVX2
* 64GB RAM
* RTX3080 10GB

## Timing
The execution times of the following lines are measured for each respective task
* **Generator:** `local_data = self.generator_obj.__getitem__(index_dataset)`
* **Inference:** `predictions_data = self.model.predict_on_batch(local_data[0])`

Each line processes a batch with `batch_size=8`. Since the times may have a large degree of variability across batches, the average time across a large number of batches is taken. The average time per batch is divided by the batch size to estimate the average process time per frame. The sample size of the batches was chosen by eye when it is determined that the average has converged to a stable value. 

# Results

First, let's look at the overall times for the generator and inference optimizations on each respective workstation to visualize the proportion of time taken for each. 

![image](https://user-images.githubusercontent.com/10882060/211322322-71b89e08-1a8d-4eec-adcd-68e160d09aa9.png)

Starting with the GPU Workstation, you can see that optimizations for the generator are more drastic than the inference. The GPU generator in red is not visible because it is several orders of magnitude smaller.  Furthermore, batch 2D slicing benefits more with larger batch sizes, so the relative performance gain is dependent on the batch size.   

![image](https://user-images.githubusercontent.com/10882060/211125339-201ac708-ae70-4852-97b6-3504f9acff0e.png)

On the CPU Workstation, a similar story emerges. The generator with batch 2D slicing becomes relatively a small proportion of the processing time compared to inference.

![image](https://user-images.githubusercontent.com/10882060/211615155-a28efc7c-e749-477f-9bc1-f129eb24ba30.png)

The generator optimizations range on the scale of orders of magnitudes and are represented by a log scale. 

![image](https://user-images.githubusercontent.com/10882060/211125606-aadb31ac-6034-4e3b-abb8-8dbb29443f77.png)

AVX512 extensions with Intel MKL ran about 35% faster. This is not as drastic as [3x performance gains reported by DeepVariant](https://google.github.io/deepvariant/posts/2019-04-30-the-power-of-building-on-an-accelerating-platform-how-deepVariant-uses-intels-avx-512-optimizations/). While running tensorflow, it's still returning warnings that AVX-512FMA is supported by not enabled. There are multiple variants of AVX-512 that may potentially be further optimized. 

Python multiprocessing being faster than tensorflow multithreading was a big surprise to me. Why would python's multiprocessing with its infamous GIL issue be faster than tensorflow's C++ backend that utilizes Intel's MKL? [This medium article addressed this](https://muditb.medium.com/speed-up-your-keras-sequence-pipeline-f5d158359f46), and claims it is due to a bottleneck in the data transfer between processes and remedies it by creating a multiprocess shared_mem object. He suspiciously doesn't show the results of his solution in the end. It might be worth trying out to see if it speeds up tensorflow's multithreading. 

![image](https://user-images.githubusercontent.com/10882060/211617750-a56a702b-5ed8-42f2-9d0e-37bb7e52a3bb.png)

The GPU generator shows a big performance gain when the transfer of the batch to GPU memory is no longer a part of the model.predict_on_batch.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +900,144 @@ class OphysGenerator(SequentialGenerator):
+    def __getitem__(self, index):
+        if self.gpu_available:
+            if self.gpu_cache_full:
+                batch_indices = self.generate_batch_indexes(index)
```
remove duplicated code block by just setting `batch_indices` and `input_indices` at top of method

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +900,144 @@ class OphysGenerator(SequentialGenerator):
+        return input_full, output_full
 
-        return data_img_input, data_img_output
+    def __index_generation__(self, index_frame):
```
Only "magic" methods should use `__ __`. please make private instead. Also method names should start with an action verb, such as `get`. Also the name for this method can be made more clear as to "what" indices are being generated such as `_get_indices_for_frame_block`

aamster commented on deepinterpolation/inferrence_collection.py:
```diff
@@ -244,18 +244,14 @@ def run(self):
-            for epoch_index, index_dataset in enumerate(tqdm(np.arange(0, self.nb_datasets, self.steps_per_epoch))):
-
+            for epoch_index, index_dataset in enumerate(tqdm(np.arange(self.nb_datasets))):
+                local_data = self.generator_obj.__getitem__(index_dataset)
```
do not call magic methods directly. Instead use 
```
local_data = self.generator_obj[index_dataset]
```

aamster commented on setup.py:
```diff
@@ -1,20 +1,29 @@
-with open("requirements.txt", "r") as f:
-    required = f.read().splitlines()
+if platform.system == 'Linux' or platform.system == 'Darwin':
+    if os.system("grep avx512 /proc/cpuinfo") == 0:
```
Please do not maintain duplicated requirements in 2 files. Just add the additional requirements in `requirements-avx512.txt`. I would also consider this an "extra" that needs to get installed separately. That can be added to `extra_requires` as 
```
extra_requires={
    'avx512': avx512_requirements
}
```
this means we'll have to check whether cpu is avx512 in the install script rather than here, which i think is fine.

mikejhuang replied:
intel-tensorflow-avx512 gets substituted for tensorflow
So it's not really an extra. I wrote a script to do the substitution with one requirements.txt. Let me know if you think there are issues with this revision.

aamster replied:
got it. Much improved, thanks.

aamster reviewed (changes requested):
In addition to the feedback, please also add tests that show that the code works as expected.
Also, all github actions checks need to pass

aamster commented on setup.py:
```diff
@@ -1,20 +1,29 @@
 setup(
     name="deepinterpolation",
-    version="0.1.5",
+    version="0.1.6",
```
this can be removed. It will be read from `__init__.py`

mikejhuang commented on deepinterpolation/cli/schemas.py:
```diff
@@ -197,6 +197,21 @@ class GeneratorSchema(argschema.schemas.DefaultSchema):
             the maximal number."
     )
 
+    gpu_cache_full = argschema.fields.Boolean(
```
There's a gpu_available conditional. It only runs this option store whole movie as tensor if gpu_available=True
Good point with mentioning it's just for OphysGenerator.

mikejhuang replied:
What's the reason to rename this with the private variable name convention?

mikejhuang commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +900,144 @@ class OphysGenerator(SequentialGenerator):
-        if self.cache_data:
-            movie_obj = self.raw_data
+        if self.gpu_available:
+            self.data_tensor = False
```
This line is used as a placeholder to check if the initial batch has been loaded yet. If it has, it only loads the next subsequent frames that are different between the new batch and the previous batch.

mikejhuang commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +900,144 @@ class OphysGenerator(SequentialGenerator):
+            self.data_tensor = False
+            if self.gpu_cache_full:
+                print("Caching full movie onto GPU")
+                self.data_tensor = tf.convert_to_tensor(
```
What do you think about the fact that this returns a different data type and the operations on this data type are different (e.g. slicing, moveaxis, expand_dims)

aamster replied:
It's fine

mikejhuang commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +900,144 @@ class OphysGenerator(SequentialGenerator):
+        return self._movie_data
+
+
+    def __gpu_cache__(self, index):
```
I don't see any point to doing this on the CPU though, especially if the full movie is already cached onto RAM.

mikejhuang commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +900,144 @@ class OphysGenerator(SequentialGenerator):
+
+            input_full = tf.gather(self.data_tensor, input_indices)
+            output_full = tf.gather(self.data_tensor, batch_indices)
+            input_full = tf.experimental.numpy.moveaxis(input_full, 1, -1)
```
tf.transpose(arr) is equivalent to swapaxes(arr, 0, -1)

The dim before moveaxis is (sample, frames, x, y). It needs to be (sample, x, y, frames)

aamster replied:
it takes a `perm` argument which gives the dimension ordering. Trying to avoid using an experimental api

mikejhuang commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +900,144 @@ class OphysGenerator(SequentialGenerator):
+                # Get indices for partial movie cache
+                start = self.pre_frame + self.pre_post_omission
+                end = start + self.batch_size
+                self.batch_indices = list(range(start,end))
```
I wanted to set this instance variable because the input_indices and batch_indices are constant across all batches when slicing from the chunked movie. If it's set in __get_item__, it would regenerate the batch_indices every time it's called, which would be unnecessary compute.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,143 @@ class OphysGenerator(SequentialGenerator):
+                        self.local_mean,
+                        self.local_std)
+        if self.gpu_cache_full:
+            if self._data_tensor is None:
```
I don't think you need this extra variable `self._data_tensor`. Can't you just convert `movie_data` to tensor and use `self.movie_data` throughout?

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,143 @@ class OphysGenerator(SequentialGenerator):
+            if self._data_tensor is None:
+                logger.info("Caching full movie onto GPU")
+                self._data_tensor = tf.convert_to_tensor(
+                    self.movie_data, dtype="float")
```
need to refer to private variable, otherwise this is recursive.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,143 @@ class OphysGenerator(SequentialGenerator):
+                        for frame_index in self.batch_indices])
+
+    @property
+    def movie_data(self) -> Union[np.ndarray, tf.Tensor]:
```
why not convert to `Tensor` always and just work with that instead of 2 different data types?

mikejhuang replied:
Actually, after testing this, the CPU processing when first converted to a tensor is much slower. Perhaps the slicing operation isn't nearly as efficient with tf.Tensor as opposed to np.array.

aamster replied:
ok. you mean that it is slow only the first time an operation is done on the tensor? I think that is a fine tradeoff to make since it simplifies the code and reduces the chance for bugs.

mikejhuang replied:
Nope, it's slower for each batch that's being processed which consists of batch generation (slicing) and inference. My test on a movie truncated to 5000 frames indicated that it takes 36 minutes as a numpy array and 62 minutes as a tf.Tensor with CPU processing.

aamster replied:
That's bad. One of the reasons why I prefer pytorch.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,143 @@ class OphysGenerator(SequentialGenerator):
-            data_img_input = movie_obj[input_index, :, :]
-            data_img_output = movie_obj[index_frame, :, :]
+    def __getitem__(self, index):
+        if not (self._gpu_available and not self.gpu_cache_full):
```
This is checking if not self._gpu_available and self.gpu_cache_full ? That should never be true?

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,143 @@ class OphysGenerator(SequentialGenerator):
+        return input_full, output_full
 
-        return data_img_input, data_img_output
+    def __index_generation(self, index_frame):
```
needs type, docstring. Use an active verb `_get_batch_indices` or similar.

aamster reviewed (changes requested):
Looks much improved! 

I think it can be improved further by using tensor both on cpu and gpu. That way we don't need separate logic for dealing with numpy array vs tensor. I still stand behind removing "gpu" from caching arguments such as "gpu_cache_full" if we use that suggestion. 

There still needs to be tests added.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+                if self.end_frame > 0 and  total_frame_per_movie > end_ind:
+                    self._movie_data = movie_obj['data'][:end_ind]
+                else:
+                    self._movie_data = movie_obj['data'][()]
```
It's clearer if you set a local variable `movie_data`, then set `self._movie_data` at the very end once you have the final value. Easier to track where the instance variable gets set that way.

mikejhuang replied:
Nope, it's slower for each batch that's being processed which consists of batch generation (slicing) and inference. My test on a movie truncated to 5000 frames indicated that it takes 36 minutes as a numpy array and 62 minutes as a tf.Tensor with CPU processing.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+                    self._movie_data = movie_obj['data'][()]
+            if self.gpu_cache_full:
+                logger.info("Caching full movie onto GPU")
+                self._data_tensor = tf.convert_to_tensor(
```
IMO it's clearer if you explicitly move the data to gpu using `with tf.device(...)`. if you just do `convert_to_tensor`, it's unclear that it's being moved to gpu.

mikejhuang replied:
I hear you. I think comments can make that clear as well. I would prefer to let tensorflow's built in logic handle which device to use, especially if there's more than one GPU device.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+        if self._gpu_available:
+            self._cached_index = None
+            self._data_tensor = None
+            self.batch_indices = None
```
make private

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+        self.cache_data = self.json_data.get("cache_data", False)
+        self.raw_data_file = self.json_data.get(
+            "train_path", self.json_data.get("movie_path"))
+        self.gpu_cache_full = self.json_data.get("gpu_cache_full", False)
```
A user with only a cpu could pass `gpu_cache_full` = True. Then you log "Caching full movie onto GPU" which isn't true. You need to make sure that the user has a gpu if `gpu_cache_full` = True. You need to validate here that there actually is a gpu if True is passed.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+        self.local_std = local_data.std()
+
+    @property
+    def movie_data(self) -> Union[np.ndarray, tf.Tensor]:
```
The return type is incorrect. You only return np array. I made a suggestion below for how to fix it.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+                    self._movie_data = movie_obj['data'][()]
+            if self.gpu_cache_full:
+                logger.info("Caching full movie onto GPU")
+                self._data_tensor = tf.convert_to_tensor(
```
I propose to set `movie_data` to the tensor. Then the return type will be correct.

mikejhuang replied:
Yeah, it's a lot slower for processing each batch on the CPU. Might be the slicing operation is less efficient with tf than np.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+        elif self._cached_index == index:
+            return self._data_tensor
+        
+        else:
```
clearer if you instead do 
```
if self._cached_index is not None:
    if self._cached_index == index-1:
        ....
    else:
        ...
else:
    ....
```

otherwise this else block could in theory be triggered unexpectedly

mikejhuang replied:
The function of the first block is to copy just the difference of frames between the previous index, so this should only be run when self._cached_index == index-1. 

If there `self._cached_index is None or self._cached_index != index-1 or self._cached_index != index, it`, it should be caching the entire batch movie instead. 

So I think the current logic is correct.

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+                        self.local_std)
+        return self._movie_data
+
+    def batch_tensor(self, index):
```
please make this `_get_batch_tensor`

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
-            self.raw_data = ( self.raw_data.astype("float") \
-                - self.local_mean ) / self.local_std
+        elif self._cached_index == index:
+            return self._data_tensor
```
I propose to rename `self._data_tensor` to `self._batch_tensor`

aamster commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
+        self._gpu_available = tf.test.is_gpu_available()
+        self._movie_data = None
+        if self._gpu_available:
+            self._cached_index = None
```
Please rename to something clearer like `_prev_batch_index`

mikejhuang replied:
It's the current batch index

aamster replied:
So is `_current_batch_index` a clearer name?

mikejhuang replied:
I set it to `_batch_tensor_index`. I wanted to specify that it specifically refers to the index for `self._batch_tensor`. I think current is implied by the name? I added a comment next to it to clarify for anyone reading the code.

aamster reviewed (changes requested):
In addition to the feedback, please
a) write tests
b) make sure all github action checks pass

mikejhuang commented:
> In addition to the feedback, please a) write tests b) make sure all github action checks pass

First addressing the comments, will write the tests now.

mikejhuang commented:
Currently, two tests are failing on the base branch `fix/faster_inference`
1. tests/cli/test_inference.py::test_integration_cli_ephys_inference_padding - I think this is due to changes in the start/stop logic. 
2. tests/test_inference_collection.py::test_mlflow_inference - has an issue importing a module from SQLAlchemy. Might need to check dependency versions and determine cross-compatibility. 
```
>   from sqlalchemy.sql.expression import _BindParamClause
E   ImportError: cannot import name '_BindParamClause' from 'sqlalchemy.sql.expression' (/home/circleci/.pyenv/versions/3.7.16/lib/python3.7/site-packages/sqlalchemy/sql/expression.py)
```

aamster reviewed (approved):
Looks good

jeromelecoq commented on deepinterpolation/generator_collection.py:
```diff
@@ -89,6 +96,21 @@ def __get_norm_parameters__(self, idx):
 
         return local_mean, local_std
 
+    def _normalize(self, arr: Union[np.ndarray, tf.Tensor], 
```
I understand this function is to avoid duplication in the OphysGenerator. 
At the moment it probably belongs in the OphysGenerator class since it is now used in any other sub-class of DeepGenerator.

jeromelecoq commented on deepinterpolation/cli/schemas.py:
```diff
@@ -197,6 +197,31 @@ class GeneratorSchema(argschema.schemas.DefaultSchema):
+        GPU at once."
+    )
+
+    normalize_cache = argschema.fields.Boolean(
```
SO these new parameters only work with OphysGenerator. I would either mention that they only apply to OphysGenerator or make them used across the others.

jeromelecoq commented on deepinterpolation/inferrence_collection.py:
```diff
@@ -244,18 +229,13 @@ def run(self):
-                        steps = local_length, max_queue_size = 10,
-                        workers = self.workers,
-                        use_multiprocessing = self.use_multiprocessing)
+                predictions_data = self.model.predict_on_batch(local_data[0])
```
So this is where I want to test whether getting rid of multiprocessing make sense. I am acknowledging that you removing this parameter (self.workers and self.use_multiprocessing) from doing anything.

jeromelecoq commented on requirements.txt:
```diff
@@ -1,4 +1,4 @@
-tensorflow
+tensorflow==2.7
```
what is the rationale for enforcing this specific version?

jeromelecoq commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +919,147 @@ class OphysGenerator(SequentialGenerator):
-            self.raw_data_file = self.json_data["train_path"]
+        self.cache_data = self.json_data.get("cache_data", False)
+        self.raw_data_file = self.json_data.get(
+            "train_path", self.json_data.get("movie_path"))
```
you are changing the parameter train_path name here.

jeromelecoq commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +919,147 @@ class OphysGenerator(SequentialGenerator):
-        movie_obj_point = h5py.File(self.raw_data_file, "r")
-        raw_data = movie_obj_point["data"]
+        self.movie_statistics_sample_size = \
+            self.json_data.get("movie_statistics_sample_size", 1000)
```
name of variable is a bit unclear "movie_statistics_sample_size", perhaps 
"movie_statistics_nbframes"

jeromelecoq commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +919,147 @@ class OphysGenerator(SequentialGenerator):
-        else:
-            self.cache_data = False
-
-        if self.cache_data:
```
This section is important. It allows to cache only a part of the movie if the entire movie is not asked for.

jeromelecoq commented on deepinterpolation/generator_collection.py:
```diff
@@ -897,112 +919,147 @@ class OphysGenerator(SequentialGenerator):
+    @property
+    def movie_data(self) -> Union[np.ndarray, tf.Tensor]:
+        if self._movie_data is None:
+            with h5py.File(self.raw_data_file, "r") as movie_obj:
```
As far as I understand, this code now forces everyone to load the entire set of frames in RAM no matter what. The generator was written to allow lazy loading so that people with minimal RAM could do this OR we could handle very very large movies. I would prefer to allow lazy loading. The previous generator was also written to repoint at a file when fetching batches.

jeromelecoq replied:
Given that the entire generator was rewritten, I have to check its behavior as this code was the result of months of optimization and edge-cases exploration. Isn't there a simpler way to add the GPU loading?

mikejhuang reviewed (commented):
Made the changes you requested. The partial caching will be a TODO.  I will start merging the two PRs if you are ok with the changes I made and don't have additional requests.
